### ISSUE TITLE
feature: extract categories and normalize category links - [merged]

### ISSUE THREAD

**Issue by appledora** [appledora/mwparserfromhtml#50]
_Merges 7-category-extraction -> main_

Closes #7

**Comment by appledora** [comment]
requested review from @martingerlach

**Comment by appledora** [comment]
added 1 commit

<ul><li>52d2db45 - update: added documentation</li></ul>

[Compare with previous version](/repos/research/html-dumps/-/merge_requests/4/diffs?diff_id=6029&start_sha=c8373312012108d9f4b2b349c9e0bf3820e22324)

**Comment by appledora** [comment]
In GitLab by @martingerlach on Jul 1, 2022, 16:43

Commented on [src/parse/article.py line 8](https://github.com/appledora/mwparserfromhtml/compare/70939fc763196ce5418072e037d3baa0751835a2..52d2db4520ba310e7de9e356997fbb0e97fd2229#diff-421f6ed8de5947c9983829ff0c5598b1L8)



Should we also remove this import as we are not using this currently (I think we discussed this in another context too?)

**Comment by appledora** [comment]
In GitLab by @martingerlach on Jul 1, 2022, 16:47

Commented on [src/parse/article.py line 85](https://github.com/appledora/mwparserfromhtml/compare/d810eda82a0e42f17effebfb2fd542418ec646f8..4ea7d5b483f7f52bb29888b3891cda9ef8be13fc#diff-421f6ed8de5947c9983829ff0c5598b1R85)



in get_wikilinks you are using regex, should we do the same here too, i.e. re.compile("mw:PageProp/Category")? what would be the difference in the output?

**Comment by appledora** [comment]
In GitLab by @martingerlach on Jul 1, 2022, 17:03

Commented on [src/parse/utils.py line 90](https://github.com/appledora/mwparserfromhtml/compare/70939fc763196ce5418072e037d3baa0751835a2..f552c4d45083c56fa055e380a264c89e4ea9c182#diff-8c15b76cf1dd83593673058e4a61741eR90)



I think this will only work in English Wikipedia. While I havent checked the formatting in the HTML-dumps, I am sure the link-title will contain the name of the category-namespace in the corresponding wiki. For example, the category [https://en.wikipedia.org/wiki/Category:Perception](Category:Perception) in English Wikipedia appears as [https://de.wikipedia.org/wiki/Kategorie:Wahrnehmung](Kategorie:Wahrnehmung) in German Wikipedia.

**Comment by appledora** [comment]
changed this line in [version 3 of the diff](/repos/research/html-dumps/-/merge_requests/4/diffs?diff_id=6122&start_sha=52d2db4520ba310e7de9e356997fbb0e97fd2229#109c416a40a3fe2b5dc233211b25f53c2750ffea_8_7)

**Comment by appledora** [comment]
added 1 commit

<ul><li>f552c4d4 - Update article.py</li></ul>

[Compare with previous version](/repos/research/html-dumps/-/merge_requests/4/diffs?diff_id=6122&start_sha=52d2db4520ba310e7de9e356997fbb0e97fd2229)

**Comment by appledora** [comment]
Yes, auto-import by my ide. Thanks for pointing this out!

**Comment by appledora** [comment]
So the idea here is that, in Wikilink, using regex helped me to also capture the interwikilinks that has the format `mw:Wikilink/interwiki`. In case of categories, i had to use a stricter pattern because there is also another `rel` value present in the html that looks something like `mw:PageProp/Categorysort...` - which we don't want for the object here. I hope I have managed to explain my thoughts here on this :3

**Comment by appledora** [comment]
This makes sense. What would you suggest here?

**Comment by appledora** [comment]
In GitLab by @martingerlach on Jul 1, 2022, 22:33

Commented on [src/parse/article.py line 85](https://github.com/appledora/mwparserfromhtml/compare/d810eda82a0e42f17effebfb2fd542418ec646f8..4ea7d5b483f7f52bb29888b3891cda9ef8be13fc#diff-421f6ed8de5947c9983829ff0c5598b1R85)



Thanks for the explanation. That makes sense if the pattern for categories is much more narrowly defined. Then looks good to me.

**Comment by appledora** [comment]
In GitLab by @geohci on Jul 1, 2022, 22:39

Commented on [src/parse/utils.py line 90](https://github.com/appledora/mwparserfromhtml/compare/70939fc763196ce5418072e037d3baa0751835a2..f552c4d45083c56fa055e380a264c89e4ea9c182#diff-8c15b76cf1dd83593673058e4a61741eR90)



i have an issue to add these namespace names to a static file somewhere that can be imported and used for namespace detection / link cleaning so probably will have to wait till that (sorry)

**Comment by appledora** [comment]
In GitLab by @geohci on Jul 1, 2022, 22:39

now that MR5 is merged, should we rebase this to clean up?

**Comment by appledora** [comment]
I actually have very little experiences with git-rebase (scared, to be honest). But after MR#5 I was thinking of pulling those changes to this branch, modify these codes accordingly and push to the same MR#4 again. Is that the same thing?

**Comment by appledora** [comment]
In GitLab by @martingerlach on Jul 1, 2022, 22:43

Commented on [src/parse/utils.py line 90](https://github.com/appledora/mwparserfromhtml/compare/70939fc763196ce5418072e037d3baa0751835a2..f552c4d45083c56fa055e380a264c89e4ea9c182#diff-8c15b76cf1dd83593673058e4a61741eR90)



For now we could:
- keep the namespace and only strip the "./" so we keep the article-title. once we have Isaac's namespace-mapping, we can pass the full article title and split into namespace and title without namespace
- strip everything before the first ":" as a naive way to strip namespace information. in this specific case, we know the namespace (category) from the annotation of the link in the HTML

**Comment by appledora** [comment]
In GitLab by @geohci on Jul 1, 2022, 22:49

thankfully can't break anything but one approach to test first is to create another local repository on your computer, checkout this merge request using the approach i mentioned in Slack, do the git-rebase and then manually verify. if all goes well, just delete the copy of the repo and do git-rebase for real. if not, debug. i'm not sure about your proposed approach (might work) but git rebase is good to learn.

**Comment by appledora** [comment]
changed this line in [version 4 of the diff](/repos/research/html-dumps/-/merge_requests/4/diffs?diff_id=6138&start_sha=f552c4d45083c56fa055e380a264c89e4ea9c182#8a8015671fab4ab01f96d6edd9a110cbc6787757_90_86)

**Comment by appledora** [comment]
added 7 commits

<ul><li>f552c4d4...d810eda8 - 2 commits from branch <code>main</code></li><li>08ba9487 - feature: extract categories and normalize category links</li><li>53c132b5 - update: added documentation</li><li>dd321cc6 - Update article.py</li><li>406605fa - update: rebased main after MR#5</li><li>2a3c267d - update: rebased main after MR#5</li></ul>

[Compare with previous version](/repos/research/html-dumps/-/merge_requests/4/diffs?diff_id=6138&start_sha=f552c4d45083c56fa055e380a264c89e4ea9c182)

**Comment by appledora** [comment]
I took the second approach here.

**Comment by appledora** [comment]
I have tried rebasing, manually corrected the conflicts and did `git push -f`. Could you verify if it's okay, @geohci ?

**Comment by appledora** [comment]
added 1 commit

<ul><li>52d9753f - update: removed  attribute from Category class</li></ul>

[Compare with previous version](/repos/research/html-dumps/-/merge_requests/4/diffs?diff_id=6140&start_sha=2a3c267d3846ce81950696b6b4aa29a7ff370dcd)

**Comment by appledora** [comment]
added 1 commit

<ul><li>7c2d8c71 - update: minor typo fix in documentation</li></ul>

[Compare with previous version](/repos/research/html-dumps/-/merge_requests/4/diffs?diff_id=6142&start_sha=52d9753f7959208d21a90beb3a1f205d43b56f3f)

**Comment by appledora** [comment]
resolved all threads

**Comment by appledora** [comment]
In GitLab by @martingerlach on Jul 4, 2022, 14:07

Commented on [src/parse/article.py line 7](https://github.com/appledora/mwparserfromhtml/compare/d810eda82a0e42f17effebfb2fd542418ec646f8..7c2d8c71be821f529ef8075607d4817f901569bc#diff-421f6ed8de5947c9983829ff0c5598b1R7)



Category should be imported only once

**Comment by appledora** [comment]
In GitLab by @martingerlach on Jul 4, 2022, 14:29

Commented on [src/parse/utils.py line 90](https://github.com/appledora/mwparserfromhtml/compare/70939fc763196ce5418072e037d3baa0751835a2..f552c4d45083c56fa055e380a264c89e4ea9c182#diff-8c15b76cf1dd83593673058e4a61741eR90)



so this essentially aims to strip the namespace-title from the article-title. I would add a short comment to that line so we know immediately what this is doing.

however, I do wonder if this approach works in general for all titles (and not just categories). Some article titles seem to contain ":", where it is part of the actual title and is not just a marker of the namespace. In this case, the normalization would introduce an error. A quick check for English Wikipedia showed there are 35k articles in namespace 0 with ":" in its titles (another 100k titles from redirects). Thus, we might want to go with the former suggestion or restrict the current approach only to type=category.

In addition, I would suggest to avoid duplicating `link.strip().replace("_", " ")` in several different lines. We might improve this later and then might miss to apply changes to both instances.

**Comment by appledora** [comment]
changed this line in [version 7 of the diff](/repos/research/html-dumps/-/merge_requests/4/diffs?diff_id=6150&start_sha=7c2d8c71be821f529ef8075607d4817f901569bc#109c416a40a3fe2b5dc233211b25f53c2750ffea_7_7)

**Comment by appledora** [comment]
added 1 commit

<ul><li>bf727afa - Update article.py</li></ul>

[Compare with previous version](/repos/research/html-dumps/-/merge_requests/4/diffs?diff_id=6150&start_sha=7c2d8c71be821f529ef8075607d4817f901569bc)

**Comment by appledora** [comment]
added 1 commit

<ul><li>a6ad3dec - Update utils.py</li></ul>

[Compare with previous version](/repos/research/html-dumps/-/merge_requests/4/diffs?diff_id=6155&start_sha=bf727afafd773bea4c9b30424c413c3b6e18288f)

**Comment by appledora** [comment]
I have currently added the comment and updated the duplicate line. Regarding the issue of links containing ":" themselves, do all links start with a namespace marker before them? e.g: "./Template:www.linksomething.com" , "./Category:www.somethingelse.com"

**Comment by appledora** [comment]
In GitLab by @martingerlach on Jul 4, 2022, 23:47

Commented on [src/parse/utils.py line 90](https://github.com/appledora/mwparserfromhtml/compare/70939fc763196ce5418072e037d3baa0751835a2..f552c4d45083c56fa055e380a264c89e4ea9c182#diff-8c15b76cf1dd83593673058e4a61741eR90)



I think that the link-title always starts with "\<namespace\>:" except if it is the main namespace (an internal link to another article). The problem comes from the latter: If the title contains ":" then normalizing the title via `link = link.split(":", 1)[1]` will split off part of the actual link-title. For example, a link to the article [Star_Trek:_Enterprise](https://en.wikipedia.org/wiki/Star_Trek:_Enterprise) would yield "_Enterprise" (which is the wrong article).

**Comment by appledora** [comment]
It seems, for now we have know other way but to use the namespace mapping. However, there might still be ambiguations, right? What if "Template: A cool anime name" is an actual internal wikilink?

**Comment by appledora** [comment]
In GitLab by @geohci on Jul 5, 2022, 02:23

Commented on [src/parse/utils.py line 90](https://github.com/appledora/mwparserfromhtml/compare/70939fc763196ce5418072e037d3baa0751835a2..f552c4d45083c56fa055e380a264c89e4ea9c182#diff-8c15b76cf1dd83593673058e4a61741eR90)



I believe there are restrictions that prevent ambiguous article names like that: https://en.wikipedia.org/wiki/Wikipedia:Naming_conventions_(technical_restrictions)#Other_problematic_characters

I put together the patch that will include all the namespace names on Wikipedia so we should be close to adding namespace detection to the Wikilink class and will create that merge request when this is merged. makes me wonder if this function should move under the Wikilink class (and Category/File classes would have their own too) rather than try to have a single function that covers all the different types?

**Comment by appledora** [comment]
In GitLab by @geohci on Jul 5, 2022, 02:23

Commented on [src/parse/article.py line 7](https://github.com/appledora/mwparserfromhtml/compare/d810eda82a0e42f17effebfb2fd542418ec646f8..a6ad3dec93596738d03bab28bcd2b42952459376#diff-421f6ed8de5947c9983829ff0c5598b1R7)



I think the ExternalLink import was accidentally removed

**Comment by appledora** [comment]
In GitLab by @geohci on Jul 5, 2022, 02:23

Commented on [src/parse/utils.py line 14](https://github.com/appledora/mwparserfromhtml/compare/d810eda82a0e42f17effebfb2fd542418ec646f8..95eb4dea87228f03505fe451981eeb9d7a8c4b15#diff-8c15b76cf1dd83593673058e4a61741eR14)



what is the purpose of this code? if it's for plaintext generation, should we instead replace it with the [html library](https://docs.python.org/3/library/html.entities.html) that [mwparserfromhell uses](https://github.com/earwig/mwparserfromhell/blob/main/src/mwparserfromhell/nodes/html_entity.py)? if that's the case, we can probably just remove and leave for a later merge request when we get to that stage. same for the htmlDecode function later in this file if it's not being used currently.

**Comment by appledora** [comment]
In GitLab by @geohci on Jul 5, 2022, 02:23

I think so...

**Comment by appledora** [comment]
Oh yeah, currently they are not being used in any function but I had used them previously as handy utils while extracting plaintexts. It is supposed to remove the `html entities` from strings e.g: &gt, &lt etc.

**Comment by appledora** [comment]
changed this line in [version 9 of the diff](/repos/research/html-dumps/-/merge_requests/4/diffs?diff_id=6157&start_sha=a6ad3dec93596738d03bab28bcd2b42952459376#109c416a40a3fe2b5dc233211b25f53c2750ffea_7_7)

**Comment by appledora** [comment]
added 1 commit

<ul><li>95eb4dea - Update article.py</li></ul>

[Compare with previous version](/repos/research/html-dumps/-/merge_requests/4/diffs?diff_id=6157&start_sha=a6ad3dec93596738d03bab28bcd2b42952459376)

**Comment by appledora** [comment]
changed this line in [version 10 of the diff](/repos/research/html-dumps/-/merge_requests/4/diffs?diff_id=6169&start_sha=95eb4dea87228f03505fe451981eeb9d7a8c4b15#8a8015671fab4ab01f96d6edd9a110cbc6787757_14_5)

**Comment by appledora** [comment]
added 1 commit

<ul><li>4ea7d5b4 - Update: cleaned utils.py</li></ul>

[Compare with previous version](/repos/research/html-dumps/-/merge_requests/4/diffs?diff_id=6169&start_sha=95eb4dea87228f03505fe451981eeb9d7a8c4b15)

**Comment by appledora** [comment]
@geohci I am little unclear about your suggestion here tbh. We can definitely write class specific title normalization method, and it would look (and read) much better too, I think. Do you suggest we use a new issue for that?

**Comment by appledora** [comment]
In GitLab by @geohci on Jul 5, 2022, 19:57

Commented on [src/parse/utils.py line 90](https://github.com/appledora/mwparserfromhtml/compare/70939fc763196ce5418072e037d3baa0751835a2..f552c4d45083c56fa055e380a264c89e4ea9c182#diff-8c15b76cf1dd83593673058e4a61741eR90)



> We can definitely write class specific title normalization method, and it would look (and read) much better too, I think. Do you suggest we use a new issue for that?

Yep, that sounds perfect!

**Comment by appledora** [comment]
In GitLab by @geohci on Jul 5, 2022, 21:14

resolved all threads

**Comment by appledora** [comment]
In GitLab by @geohci on Jul 5, 2022, 21:15

mentioned in commit fa34018558a85fd56506cebf028ec94a813f9557
